src/rgw/rgw_sync_policy.h

```diff
@@ -354,14 +354,22 @@ struct rgw_sync_pipe_params {
   }
 
   void decode(bufferlist::const_iterator& bl) {
-    DECODE_START(1, bl);
+    DECODE_START(2, bl);
     decode(source, bl);
     decode(dest, bl);
     decode(priority, bl);
     uint8_t m;
     decode(m, bl);
     mode = (Mode)m;
-    decode(user, bl);
+    if (struct_v >= 2) {
+      decode(user, bl);
+    } else {
+      rgw_user old_user;
+      decode(old_user, bl);
+      if (!old_user.empty()) {
+        user = old_user;
```

Comment:
if the v2 encoding changes `rgw_user` to `optional<rgw_user>`,  v1 decoders will fail when they try to decode that. this could lead to errors during rolling upgrade

since your v1 decode has special handling for the `old_user.empty()` case, maybe we should just encode an empty `rgw_user` when the `optional<rgw_user>` is empty? then we wouldn't need to bump the version

Reply:
was just curious to know, what if we leave the current `decode()` as it is, and keep `encode()` one version below and modify the encode function to write empty as you suggested when it's nullopt, and then we eventually in next releases (something like T+2) bump the encode version to the same as decode and encode std::optional? would that open the case where we could eventually get rid of this kinda like hack?
I mean, perhaps we would have minimum to maximum upgrade support from version to version, so we could say that from something like T+2, we can start encoding the actual std::optional, as we can assume now that all versions understand how to decode that kind.

Reply:
i think i understand your suggestion, but i'm not sure that's worth optimizing for. if we reserve a v2 for the T+2 release, what if we need to add something else in the meantime? i know we're also interested in extending this for role support too